### PR TITLE
S2743: Add support for record struct

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/StaticFieldInGenericClass.cs [hunk 1]
```diff
@@ -40,41 +41,41 @@ public sealed class StaticFieldInGenericClass : SonarDiagnosticAnalyzer
...
         protected override void Initialize(SonarAnalysisContext context) =>
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
                 {
                     var typeDeclaration = (TypeDeclarationSyntax)c.Node;
-                    if (c.ContainingSymbol.Kind != SymbolKind.NamedType
+                    if (c.IsRedundantPositionalRecordContext()
                         || typeDeclaration.TypeParameterList == null
                         || typeDeclaration.TypeParameterList.Parameters.Count < 1)
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/StaticFieldInGenericClass.cs:49]:
```suggestion
                        || c.Node is not TypeDeclarationSyntax { TypeParameterList: { Parameters: { Count: > 0 } } } typeDeclaration)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/StaticFieldInGenericClass.cs [hunk 2]
```diff
@@ -40,41 +41,41 @@ public sealed class StaticFieldInGenericClass : SonarDiagnosticAnalyzer
...
                     }
                     var typeParameterNames = typeDeclaration.TypeParameterList.Parameters.Select(x => x.Identifier.ToString()).ToArray();
-                    var fields = typeDeclaration.Members.OfType<FieldDeclarationSyntax>().Where(f => f.Modifiers.Any(SyntaxKind.StaticKeyword));
-                    foreach (var field in fields.Where(field => !HasGenericType(field.Declaration.Type, typeParameterNames, c)))
+                    var fields = typeDeclaration.Members.OfType<FieldDeclarationSyntax>().Where(x => x.Modifiers.Any(SyntaxKind.StaticKeyword));
+                    foreach (var field in fields.Where(x => !HasGenericType(c, x.Declaration.Type, typeParameterNames)))
                     {
-                        field.Declaration.Variables.ToList().ForEach(variable => CheckMember(variable, variable.Identifier.GetLocation(), typeParameterNames, c));
+                        field.Declaration.Variables.ToList().ForEach(variable => CheckMember(c, variable, variable.Identifier.GetLocation(), typeParameterNames));
                     }
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/StaticFieldInGenericClass.cs:58]: Can be simplified:
```cs
                    var variables = typeDeclaration.Members
                        .OfType<FieldDeclarationSyntax>()
                        .Where(x => x.Modifiers.Any(SyntaxKind.StaticKeyword) && !HasGenericType(c, x.Declaration.Type, typeParameterNames))
                        .SelectMany(x => x.Declaration.Variables);
                    foreach (var variable in variables)
                    {
                        CheckMember(c, variable, variable.Identifier.GetLocation(), typeParameterNames);
                    }
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/StaticFieldInGenericClass.cs [hunk 3]
```diff
@@ -40,41 +41,41 @@ public sealed class StaticFieldInGenericClass : SonarDiagnosticAnalyzer
...
                     {
-                        CheckMember(property, property.Identifier.GetLocation(), typeParameterNames, c);
+                        CheckMember(c, property, property.Identifier.GetLocation(), typeParameterNames);
                     }
                 },
                 SyntaxKind.ClassDeclaration,
                 SyntaxKind.InterfaceDeclaration,
                 SyntaxKindEx.RecordClassDeclaration,
+                SyntaxKindEx.RecordStructDeclaration,
                 SyntaxKind.StructDeclaration);
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/StaticFieldInGenericClass.cs:68]: Ordering?

**Review by martin-strecker-sonarsource — APPROVED** [review]
That's a nice rule. Just some minor suggestions.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/StaticFieldInGenericClass.cs (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/StaticFieldInGenericClass.cs:49]: @csaba-sagi-sonarsource That's one of the places where C#10 would remove some clutter:
```cs
// C# 9
c.Node is not TypeDeclarationSyntax { TypeParameterList: { Parameters: { Count: > 0 } } } typeDeclaration)
// C# 10
c.Node is not TypeDeclarationSyntax { TypeParameterList.Parameters.Count: > 0  } typeDeclaration)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/StaticFieldInGenericClass.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/StaticFieldInGenericClass.cs:58]: Nice

**Pull request merged**